### ISSUE TITLE
Possible bug or documentation issue: beta in denstream

### ISSUE THREAD

**Issue by vsoch** [online-ml/river#874]
So I was looking to control my clustering and the beta parameter:

![image](https://user-images.githubusercontent.com/814322/157983589-689b0577-568f-43c4-ba1c-43fdcf8f7e0d.png)

I noticed that it defaults to 5 and in the example (below) it's >> 1 (1.01) but then the docs say it should be between 0 and up to 1? :thinking: 

![image](https://user-images.githubusercontent.com/814322/157983825-3ce8f252-121f-426c-ade3-d7fcafa25a93.png)

I was then trying to figure out the difference between beta and mu but they sound like the same thing in the docs!

![image](https://user-images.githubusercontent.com/814322/157983899-b6b4ea39-ce74-4300-9eea-6ae557086021.png)

So I looked at the code, and it looks like mu is only used separately from beta (usually the two are multiplied)  [here](https://github.com/online-ml/river/blob/395a61ab759938bbaea035a6f944053f509965d5/river/cluster/denstream.py#L230-L243) and [here](https://github.com/online-ml/river/blob/395a61ab759938bbaea035a6f944053f509965d5/river/cluster/denstream.py#L295-L304) so I think it has to do with the threshold of deciding if something is reachable (temporally and based on distance between clusters) and then deciding if a neighborhood should be turned into a microcluster?

Sorry I'm new at this algorithm, I was looking through the [paper](https://archive.siam.org/meetings/sdm06/proceedings/030caof.pdf) and on page 3 it does mention that the beta range should be between 0 and 1:

![image](https://user-images.githubusercontent.com/814322/157985485-2ed70b9f-dbf8-4379-8033-e3c76ce9dbff.png)

Then for mu, on page 4 in equation 4.1: 

![image](https://user-images.githubusercontent.com/814322/157985562-7230a977-b9a3-45f4-b7ce-267e06d3b053.png)

it's log (beta*mu/beta*mu -1) but a log can't be defined on a negative number, so that tells us that beta * mu has to be greater than or equal to 1, beta is positive so mu has to be bigger than 1/beta. In the example above, for example, mu would need to be greater than 0.999. Here are my notes:

```
In Definition 3.4: 0 < β ≤ 1

In Eq. 4.1 (4.1) Tp = d 1 λ log( βµ /(βµ − 1) )e,

Log(x) is defined for x > 0.

So βµ − 1 > 0
βµ >= 1
µ >= 1/β
```

So something is wrong in the docs (and let's discuss here!) My suggestions for working on this issue are:

1. clarify the difference in the algorithm's use of beta vs mu
3. clarify the appropriate ranges for each of beta and mu, assert this in the code
4. update the docs! :)

I'm happy to help with any of the above - let's discuss and I can open a PR (and I'm okay if someone else jumps on it too!)

-----------------------

Update: conceptually maybe:

- mu: the average number of points we'd expect in each cluster by the end (not a micro cluster) perhaps that's [min_samples here](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html)?
- epsilon: roughly the size (not in members but in terms of distance) of clusters "the size we'd expect to be able to cover the clusters". E.g., if we have a funny shaped cluster, this is the number of balls we need to cover without the ball extending into another cluster - see page 333 of the paper - the thing that looks like a worm). I think this would be the [eps](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html) here. 

So those two above are part of the offline part? And then beta is used in the online part, and that takes into account the temporal aspect, etc.

**Comment by MaxHalford** [comment]
Good research V! @hoanganhngo610 I'll let you join the arena once again if you don't mind :)

**Comment by hoanganhngo610** [comment]
No problem at all @MaxHalford, I will try to have this setlled down ASAP. @vsoch Once again, I am extremely impressed by your deep understanding and careful look of clustering algorithms and of this one in particular!

**Comment by vsoch** [comment]
haha you are too kind - I’m happy to help :)

**Comment by hoanganhngo610** [comment]
First of all, once again, thank you so much @vsoch for pointing the errors out. 

So basically, in DenStream, we have four parameters: `epsilon`, `beta`, `mu` and the decaying factor `lambda`. We do add in one parameter called `n_samples_init` and the number of points to be used to initialize the algorithm, since without these initalizing points, we may not get any cluster generated for predicting.

- In the online part, there are two steps, one is to merge a new point `p` into the model, the other is to check for weights of p-micro-clusters and o-micro-clusters after every time step of `T_p`. 
- In the offline part, a DBSCAN algorithm with two parameters involved `epsilon` and `mu` is built. As we are trying to minimize the depedency of `River` onto other Python packages, we decided to write the DBSCAN algorithm on our own.

From the paper, you can also easily see that `epsilon` is used as thresholds for radius of clusters, while `beta` and `mu` deals with their weights. `beta` only involves in the online part, while `mu` involves in both the online and offline part.

Now, regarding the ranges of `beta` and `mu`, I do want to discuss each point as follows:

- Regarding the range of `beta`, I agree that we do need an assertion within the code for this. Will update ASAP!
- Regarding the range of `beta * mu`, as you mentioned, it should be greater than 1. However, since we are already using the `math.log` command, I think asserting the range of `beta * mu` to be greater than 1 should not be necessary.
- Regarding the range of `mu`, since `beta` can take any value between 0 and 1, this also means that `mu` can range anywhere within the range (1, `inf`). However, as said above, I think this will only be necessary as a part of the docstring, and we do not need to do an assertion on this one.

Do you find this answer satisfactory? If not, please feel free to elaborate more on, and we can finalize a way to create a PR to solve this bug in the end!

P/S: This error may come from the fact that DenStream has been going through one round of debugging in the past, which makes the Docstring not consistent between times. My sincere apology for this!

**Comment by vsoch** [comment]
Yes, that is satisfactory! So I think all of the details you provide above should be in the docs - specifically on the page where the person is reading about the parameters and the algorithm. And I think it's important to note that we are using a modified algorithm (maybe say "inspired by" this paper?) Otherwise the reader will (as I did) go and look at the paper and try to compare things 1:1 in the case of a bug or confusion, etc.

Okay for your points! 

> Regarding the range of beta, I agree that we do need an assertion within the code for this. Will update ASAP!

Perfecto!

Regarding the range of beta * mu, as you mentioned, it should be greater than 1. However, since we are already using the math.log command, I think asserting the range of beta * mu to be greater than 1 should not be necessary.

Also agree. For this point I think the docstring just needs to be updated so that the correct range is clear

>  Regarding the range of mu, since beta can take any value between 0 and 1, this also means that mu can range anywhere within the range (1, inf). However, as said above, I think this will only be necessary as a part of the docstring, and we do not need to do an assertion on this one.

Hmm. so I think we could be more specific. Indeed mu needs to be within (1, inf), but I think we might be more specific and say (1/beta, inf)? For example if beta = 0.5, mu = 1.5 => beta * mu = 0.75 < 1.
> P/S: This error may come from the fact that DenStream has been going through one round of debugging in the past, which makes the Docstring not consistent between times. My sincere apology for this!

haha it's okay! That's what the open source eyes are here for :eyes:

**Comment by hoanganhngo610** [comment]
> Also agree. For this point I think the docstring just needs to be updated so that the correct range is clear

Agree, will do!

> Hmm. so I think we could be more specific. Indeed mu needs to be within (1, inf), but I think we might be more specific and say (1/beta, inf)? For example if beta = 0.5, mu = 1.5 => beta * mu = 0.75 < 1.

Once again, totally agree! Will specify it as `(1/beta, inf)`. 

> And I think it's important to note that we are using a modified algorithm (maybe say "inspired by" this paper?) Otherwise the reader will (as I did) go and look at the paper and try to compare things 1:1 in the case of a bug or confusion, etc.

For this, I don't think that we actually do anything to modify the algorithm too much. Every basic element is still there, just that we add in the number to specify the number of initialized data points to facilitate our `predict_one()` command. However, will explicitly note this in the docstring!

**Comment by hoanganhngo610** [comment]
This issue is solved with the PR #887.
